### ISSUE TITLE
Is there a reason my model isn't stopping generation even after producing a complete response?

### ISSUE THREAD

**Issue by yunzhongfuchen** [huggingface/nanoVLM#197]
In version 0.1, unlike the official checkpoint, which stops naturally after generating a complete response , my fine-tuned model continues generating text until it hits the `max_new_tokens` limit — often producing redundant, repetitive, or meaningless content (e.g., random names, dates, journal titles).

I suspect this is related to EOS token handling during training or inference. Could anyone help me understand what might be going wrong?

No code or configuration was modified during training — I simply ran python train.py as-is. The only change I made was to the inference script, where I adjusted the max_new_tokens parameter. Below are the output results using the official pretrained checkpoint and my own trained checkpoint, respectively:

root@dsw-659799-565c88f96b-6txt4:/mnt/workspace/nanoVLM-0.1# python generate.py 
Using device: cuda
Loading weights from: lusxvr/nanoVLM-222M

Input:
  What is this? 

Outputs:
  >> Generation 1:  This is a cat sitting on the floor. I think this is a cat sat facing towards the left side. This is a seed of the awaited fruit. This looks like an object in the flower pot. At the top right corner of the image there is another cat sitting on the floor. This is the ground. The background looks slightly blurred.
  >> Generation 2:  In the middle of the picture we can see there is a cream color of cat. We can see a white and black color combination. In the background we can see the path. At the top there is a plant which is in yellow color. On the right side of the picture it looks like a tree. The background of the picture is blur.
  >> Generation 3:  In this image I can see the cat which is in brown and white color on the cream color object. I can see the cat is in front of the black color object and the cat is in the middle of the object. At the top of the image I can see the plants and flowers and I can see another cat is right to the cat.
  >> Generation 4:  The picture is taken in a garden. In the background we can see a green garden. In the center of the picture we can see a cat which is in black and brown color. We can see that the cat is sitting on the ground. Towards the left there is a small plant on the ground. The background of the picture is very blurry.
  >> Generation 5:  In the middle of the picture, we see a cat in black and brown color. Beside that, we see flowers and other greenery. In the background, it is white in color and it is blurred and white in color. This picture is a bit dark. There is a disposal bottle in the middle of the picture and there is a text.

root@dsw-659799-565c88f96b-6txt4:/mnt/workspace/nanoVLM-0.1# python generate.py --checkpoint checkpoints/nanoVLM-222M
Using device: cuda
Loading weights from: checkpoints/nanoVLM-222M

Input:
  What is this? 

Outputs:
  >> Generation 1: This picture is of outside. In the center there is a black and brown color cat seems to be sitting on the ground. At the top there is a man pointing towards the object which seems to be the object seems to be a garbage box, there is a black color object seems to be a car and in the background there is an ever green background.

Tancred Childress  take burgers for Subway
Tancred Childress, a Texas bride groom pianist is signed by Subway ago.[Cavemanhof: Escapist] Paul McCartney and Tancred Childress on AKA – the party ayonteau

The cavemanaboutmecomeme.comitor diy 39-year-old Walker Engled pro-angle, Paul McCartney star alliance coaxdes C land from Down Under (or the Pampardes atabouts). In true early Cial teams style, two complete lives are depicted on each page.

But on the whole, this is an album stars free, and the originals are always priced pretty high. There's a lot of sunlight on the album.Bike in Mine Reader's Book Dual-Party Calendar

Ask any vintage car or vintage bike aficionado and they will tell you the record for most bike/vehicle mix sales is a 60-over 1. But only. . . Not.The Mississippi Theatre Productions - Italian, French and American (IMT) production was recorded internationally. A Miamite piano piece was performed live from 1995 to 1999 by the Chicago Drama Festival.
Cellar sponge, Cheese, Angus, Field Trip and Architecture were among the works uplifted. CATL family history was recorded.Catalog Spotted Nine Montana River Outfitters!

In this picture we can see a lady in the store with ACTION made up of her hair... MOST OFERS. And here is the buffalo buffalo.The Benefits of Bertolt Baker And Direct Album Cover Singer / vinyl: France / Japan Indie Album Art One-Dimensional - 6" - Volume One Volume One
Bertolt Baker CD by Slim Collection.Great Lakes Battlefield Park (Beltane Press.com) (German Edition) by Slim Collection presents a hardcover record of Bertolt Brecht's "The Bra
...

**Comment by yunzhongfuchen** [comment]
I've identified the issue. During data preprocessing, if only a single eos_token is appended to the end of the answer, the model may continue generating meaningless tokens even after outputting an eos_token during inference. To address this, I now append **three** eos_tokens at the end of each answer:

answer = item['answer'] + self.tokenizer.eos_token * 3

This trains the model to recognize a sequence of consecutive eos_tokens as a strong termination signal, enabling it to stop generation more reliably and avoid producing redundant or irrelevant outputs.

**Comment by aidansmyth95** [comment]
I did some further debugging. While this might help improve the symptom you described above, it actually isn't the solution. The issue is actually in the eos_token_id for the HuggingFaceTB/cosmo2-tokenizer. Let me explain.

So I assume, like me, you were using the config's setting for lm_eos_token_id: int = 0. However, when you look at the tokenizer, you find that the token ID is actually "endoftext" and not "im_end" (token ID 2).

Once I change the experiment to lm_eos_token_id: int = 2, it seems to converge for MMStar much better, peaking at about 35% accuracy. Give it a try and see, I reckon it fixes the issue.
